### DISCUSSION TITLE
Google ArtifactRegistry auth not working

### DISCUSSION THREAD

**Discussion by roman-supy-io** [renovatebot/renovate#16508] [Request Help]
I'm trying to set up private Google ArtifactRegistry repository integration. 
It has 2 types of auth:
- short living token (1h)
- long living service account (no time limit)

I am following the steps at [https://cloud.google.com/artifact-registry/docs/nodejs/authentication#auth-password](https://cloud.google.com/artifact-registry/docs/nodejs/authentication#auth-password), the values were inserted at the config below.

The latter I'm trying to use with Renovate App.
This is my `renovate.json`:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base"],
  "hostRules": [
    {
      "matchHost": "https://europe-west4-npm.pkg.dev/<project id>/npm/",
      "username": "_json_key_base64",
      "encrypted": {
        "password": "<renovate encoded password>"
      }
    }
  ],
  "rangeStrategy": "pin"
}
```
and this is the `.npmrc`, that's checked into repository:
```
@alias:registry=https://europe-west4-npm.pkg.dev/<project id>/npm/
@anotheralias:registry=https://europe-west4-npm.pkg.dev/<project id>/npm/
//europe-west4-npm.pkg.dev/<project id>/npm/:always-auth=true
```
As a result, I'm getting the `403` error.

The same secrets work as intended with the `.npmrc`, if i put them like this:
```
@alias:registry=https://europe-west4-npm.pkg.dev/<project id>/npm/
@anotheralias:registry=https://europe-west4-npm.pkg.dev/<project id>/npm/
//europe-west4-npm.pkg.dev/<project id>/npm/:always-auth=true
//europe-west4-npm.pkg.dev/<project id>/npm/:username=_json_key_base64
//europe-west4-npm.pkg.dev/<project id>/npm/:_password=<base 64 encoded service account>
```

What am I doing wrong?

P.S. I was also trying to use the `<renovate encoded password>` as a `token`, however, it doesn't seem to work of course.

**Comment by viceice** [comment]
please format code blocks

**Reply by roman-supy-io** [reply]
@viceice could you please advise?

**Comment by rarkins** [comment]
Ensure that you have taken the `<base 64 encoded service account>` string and base64 decoded it to get the "actual" password. Now encrypt *that* decoded string using the Renovate app's encryption page. If you instead encrypted the base64 encoded password that's in the npmrc, it won't work.

**Reply by roman-supy-io** [reply]
@rarkins i did that of course, what's in `.npmrc` is raw password generated by google, what's in `renovate.json` is the password encoded with RenovateApp's public key.

**Reply by roman-supy-io** [reply]
that's why it's wrapped in `"encrypted"`

**Reply by rarkins** [reply]
Per my understanding:
- passwords in npmrc are not raw, they are base 64 encoded passwords
- If you copy pasted that same value into Renovate's encryption form then it is pretty much guaranteed not work

**Reply by rarkins** [reply]
I think you can probably verify what I'm saying by looking at this source code: https://github.com/renovatebot/renovate/blob/e10aa6aa5a47056471b1cfa0d58cbc0ba16083e7/lib/modules/manager/npm/post-update/rules.ts#L40-L47

**Reply by roman-supy-io** [reply]
@rarkins I probably was not clear enough, in describing things above, by `raw` I meant, it's not encrypted by Mend's public key.
So google generated password is already base64-encoded and in `.npmrc` I use it as is, and it works of course as per their tutorial.
In `renovate.json` on the other hand, I encrypt the above-mentioned base64-encoded password with Mend's public key, and as per docs, it's also correct.

Now the problem, I think, is in how RenovateApp sends these credentials to the Artifact Registry, because if you check the `.npmrc`, the key for Password is actually `_password` and in the output logs I see the `authorization` header is being sent probably containing this secret (not sure because it's redacted with ****), while `username` and `password` fields are empty.
I'm not sure how exactly npm sends these credentials under the hood, but i suppose, RenovateApp is doing this in a different way probably.

**Reply by rarkins** [reply]
Please re read what I wrote the last two times. I think you are misunderstanding me and it's the most likely cause of your problem. Whatever is in your npmrc needs base64 decoding before it's encrypted.

**Reply by roman-supy-io** [reply]
@rarkins you are right, for some reason the string produced by google is base64 encoded **twice**(!)

**Comment by PhilipAbed** [comment]
does renovate know how to replace this project id? 
`https://europe-west4-npm.pkg.dev/<project id>/npm/`

**Reply by rarkins** [reply]
I believe that's just a redaction and he's using the actual project ID in host rules and npmrc

**Reply by roman-supy-io** [reply]
@PhilipAbed @rarkins yes, i don't want to expose my company's private data, that's why it's redacted.

**Reply by PhilipAbed** [reply]
:D just checking thx

**Comment by roman-supy-io** [comment]
@rarkins @PhilipAbed I've got one more problem with `.npmrc`. In case there are aliases that overlap, RenovateApp is not able to identify the correct registry for one of them.

`.npmrc`:
```
@alias:registry=https://europe-west4-npm.pkg.dev/<project id>/npm/
@alias.api:registry=https://europe-west4-npm.pkg.dev/<project id>/npm/
.....
```

When looking into RenovateApp logs:
`@alias.api` is pointing to the `https://europe-west4-npm.pkg.dev/<project id>/<package_name>` (correct)
while for some reason
`@alias` is pointing to the `https://registry.npmjs.org/<package_name>...` (incorrect)

Could anyone please advise?

**Reply by roman-supy-io** [reply]
this doesn't happen when I have just one of them defined.
